rfcs/019 - Datastore Compute RFC.md

```diff
@@ -0,0 +1,202 @@
+
+Sometimes this computation would also need to mutate the remotely-stored data, and sometimes it would need to reference and/or mutate the data defined outside of the closure. That means the most general purpose opcode would be a `mutate_and_map` opcode that takes a datastore namespace-key pair and a closure function that is given a mutable reference to the data and can both return a value and mutate outer scope values.
+
+But while the most general, this would have poor performance in several circumstances:
+
+1. If it is given a mutable reference to the datastore key, the only way to safely trigger multiple of these operations at the same time is if it locks the key for the duration of this remote code execution.
+2. Current closure scope handling does not know (or care) which outer-scope values are being mutated by the closure (or closure's closure, etc), so if the outer scope has a massive amount of data, even if not by this closure, it would have to copy back and forth.
+3. If the remote execution is a "fire-and-forget" mutation on the remote data, the original node will pause execution until it is completed on the other side anyways. (This can be partially worked around through an intermediate data structure that can be more quickly updated, but not eliminated with the proposed opcode.
+
+The first point can be avoided if you can specify whether or not you want a mutable reference to the remote data, the second point if you can specify whether or not you want to use closure semantics or "pure function" semantics (or something in-between), and the third can be avoided if you can specify whether or not you want a return value. With a sufficiently-advanced compiler you could have that all automatically inferred from the way it is being used (is the output assigned to a variable, used in a method chain, operator, etc, or not; is the input argument to the function ever mutated, are any outer-scope variables accessed) but for the first pass, it is proposed that we tackle this with separate explicit opcodes. This is useful even if/when automatic performance improvements to the most-general opcode are handled by the compiler as:
+
+1. It could just emit those well-defined, tighter-bound opcodes and this logic doesn't need to live in the AVM at runtime.
+2. Developers could explicitly opt-in to tighter constraints in "hotter" paths of distributed computing to prevent other developers (or their future self) from accidentally ratcheting down the runtime performance.
+
+Stepping up to the higher-level API design:
+
+```ln
+const baz = ns('someNamespace').ref('foo').run(fn (foo: string) = foo.length());
+```
+
+```ln
+let bar = 'bar';
+const baz = ns('someNamespace').ref('foo').closure(fn (foo: string) {
+  bar = 'foobar: ' + foo + bar;
+  return foo.length();
+});
+```
+
+```ln
+const bar = 'bar';
+const baz = ns('someNamespace').ref('foo').with(bar).run(fn (bar: string, foo: string) = #bar + #foo);
+```
```

Comment:
I had mixed feelings having the distinction between `run` and `closure` but having it explicit define is easier to process and less ambiguous to the user so nothing here, just a comment :)